emacs.d/configuration.org
#+TITLE: Emacs Config
#+AUTHOR: Evan Miller
#+EMAIL: nil
#+OPTIONS: toc:nil num:nil

* TODO testing 
- [ ] Testing

* User Experience
** General

#+BEGIN_SRC emacs-lisp
  (setq inhibit-startup-message t)
  (setq ring-bell-function 'ignore)
  (setq make-backup-files nil)
  (menu-bar-mode 0)
  (tool-bar-mode 0)
  (scroll-bar-mode -1)
  (blink-cursor-mode -1)
  (set-default 'cursor-type 'box)
  (size-indication-mode t)
  (global-visual-line-mode t)
  ;(toggle-truncate-lines t)
  ;(global-hl-line-mode +1)
  ;(set-face-background 'hl-line "#F13333")
#+END_SRC

** Use Fancy Lambdas

#+BEGIN_SRC emacs-lisp
  (global-prettify-symbols-mode t)
#+END_SRC

* Programming Environments
#+BEGIN_SRC emacs-lisp
  (setq-default tab-width 2)
#+END_SRC

** utf-8
#+BEGIN_SRC emacs-lisp
  (set-language-environment "UTF-8")
  (set-default-coding-systems 'utf-8)
#+END_SRC

* Org Mode
** Hooks

#+BEGIN_SRC emacs-lisp
  (add-hook 'org-mode-hook #'org-bullets-mode)
  (add-hook 'org-mode-hook #'org-indent-mode)
#+END_SRC

** Quickly Insert a Block of E-Lisp:

#+BEGIN_SRC emacs-lisp
  (add-to-list 'org-structure-template-alist
	       '("el" "#+BEGIN_SRC emacs-lisp\n?\n#+END_SRC"))
#+END_SRC

** Babel

#+BEGIN_SRC emacs-lisp
  (org-babel-do-load-languages
   'org-babel-load-languages
   '((emacs-lisp . t)
     (python . t)
     (dot . t)
     (gnuplot . t)))
#+END_SRC

** Ellipsis, Fonts, and Tabs

#+BEGIN_SRC emacs-lisp
  (setq org-ellipsis "⤵")
  (setq org-src-tab-acts-natively t)
  (setq org-src-fontify-natively t)
#+END_SRC

** Automatically Evaluate E-Lisp Code Blocks

#+BEGIN_SRC emacs-lisp
  (setq org-confirm-babel-evaluate nil)
#+END_SRC

* Ido Mode

#+BEGIN_SRC emacs-lisp
(setq ido-enable-flex-matching nil)
(setq ido-create-new-buffer 'always)
(setq ido-everywhere t)
(setq ido-vertical-define-keys 'C-n-and-C-p-only)
(ido-mode 1)
(ido-vertical-mode 1)
;(setq ido-file-extensions-order '(".org" ".txt" ".py" ".emacs" ".xml" ".el" ".ini" ".cfg" ".cnf"))

(defun bind-ido-keys ()
  (define-key ido-completion-map (kbd "C-l") 'ido-up-directory))
(add-hook 'ido-setup-hook #'bind-ido-keys)
;https://emacs.stackexchange.com/questions/3729/how-do-i-bind-keys-in-ido
#+END_SRC

* Package: Scrolling: Golden Ratio

#+BEGIN_SRC emacs-lisp
(require 'golden-ratio-scroll-screen)
(global-set-key (kbd "M-v") 'golden-ratio-scroll-screen-down)
(global-set-key (kbd "C-v") 'golden-ratio-scroll-screen-up)
#+END_SRC

* Magit

#+BEGIN_SRC
(global-set-key (kbd "C-x g") 'magit-status)
#+END_SRC

* Keybinding: Evaluate Buffer

#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-x C-b") 'eval-buffer)
#+END_SRC

* Helm

#+BEGIN_SRC emacs-lisp
(require 'helm-config)
;(helm-mode 1)

;(define-key helm-map (kbd "<tab>") 'helm-execute-persistent-action)

(global-set-key (kbd "M-x") 'helm-M-x)
;(global-set-key (kbd "C-x b") 'helm-buffers-list)
;(global-set-key (kbd "C-x r b") 'helm-bookmarks)
;(global-set-key (kbd "M-y") 'helm-show-kill-ring)
;(global-set-key (kbd "C-x C-f") 'helm-find-files)
;(global-set-key (kbd "C-x p") 'package-list-packages)
#+END_SRC

* Fonts

#+BEGIN_SRC emacs-lisp
;(set-default-font "Liberation Mono 18")
;(set-default-font "Fira Code 18")
(set-default-font "Hack Nerd Font 18")
#+END_SRC

* Themes

#+BEGIN_SRC emacs-lisp
  (setq custom-safe-themes t)
  (add-to-list 'custom-theme-load-path "~/.emacs.d/themes")
  (add-hook 'after-init-hook (lambda () (load-theme 'xresources)))
#+END_SRC

* Function: Insert Line Before

#+BEGIN_SRC emacs-lisp
(defun my/insert-line-before (times)
  "Inserts a newline(s) above the line containing the cursor."
  (interactive "p")
  (save-excursion
    (move-beginning-of-line 1)
    (newline)))

(global-set-key (kbd "C-S-o")
		'my/insert-line-before) 
#+END_SRC

* Package Config

#+BEGIN_SRC emacs-lisp
(setq package-archives
     '(("gnu" . "https://elpa.gnu.org/packages/")
       ("marmalade" . "https://marmalade-repo.org/packages/")
       ("melpa" . "https://melpa.org/packages/")))
#+END_SRC

* Custom Set Variables: DNE

(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(ansi-color-faces-vector
   [default default default italic underline success warning error])
 '(ansi-color-names-vector
   ["#191C25" "#C16069" "#A2BF8A" "#ECCC87" "#80A0C2" "#B58DAE" "#86C0D1" "#F0F4FC"])
 '(custom-enabled-themes nil)
 '(custom-safe-themes
   (quote
    ("4bbe3471ce9dfeef921417b9c1f11aacfee76790d1bc7e97986a552f37ada78b" "4d8c2744598d3b7a2bae71724fb22afbdd98bb124ad12b972fdde232ce6382ff" "714ccfcbfa15f944b077919bac4d83ecb6732f5b3b235b1066fdc81857f95150" "340eb0fc650fb288f2c138532298b34179e96ced4e21c47fedd075a8db1c1bd7" "11b09c99fbebdea84cae686192b8776beddb1796bd68574c6e763292e9ba6c54" "2612fbf211ac38c6f936fa2a8fb9e40e648e3cc481d9ef4670b63f30d0d1d782" "9319d1202111cb804f87b06982b1f01b0708ef2753a4d174fbc74429d5b4973a" "575dce634ffaf4510c93e4a50527448b843ad1f63e42bc7ae3a98186bf096153" "fe54aad86c1b38d0aae9b3d150e25030f77966287fc307c855c74a49df374825" "f7d175b582b1b7b1a5212838c98301f7968bfa190ddcb194e6e171aeb7e808b0" "bda7f790197a06eff9b0102ca86475fd4da553148809043a1ed1401fcae09b46" "0a118fe0a26858fdf29b0030382bb6a5c1b035285287b957f204a2e79134d610" "a3d7391f224603f2ec6da7948e188ea4abd9d9382ff3a9dbfae6b49279513f44" "b02a7ffc2a1b6702f0d9576fa03c46c7f63e33f2fe1162169b598e0540e257df" "47f5e1c83ba3a3ebc0acd3435beaff2c7736dddb0afc07a7e9ab99658984f4f9" "eef31981e5c59e2f17ab5a0103d8f6b0ca98e003343115ff43db7ab9a7ebbe01" "8bdd67158dcc1f792e937c093c9af97411ae129b7f0c299ea728959cec75fc4e" "749f40651187dfe531dbd8c42433ac277f6c534cdc520700b8cc88dd77dcee3f" "ffdccc09caa8a64a01d33a88b4a675608506196d933fe5bb194695d68fbb4c78" "3ebb09e6b46ec7901f7b36f3b40e2a3788c26aa07d62c7f1f41026c6eda19b1d" "aa9eb5bfeb21413e8505296c3b0ac1eb0f4be42ca5d9a99c0b988aefd4713ea1" "54e7b12a03fbab724b8202af8f64908ce1ae7fefc45d1373a8a6d15d31b33f77" "99c87b9f5233fc7697028a564ddb9e2007f351464b7097987a64d3fe2ef89c3c" "3186188aff4d89145109220e37c7b5f34ee8e408f9cd3500b8bbc16db23bd1a0" "f598f80fbdf4ed5f5a5da29dcf50f73e3f847ca5e545edf1ffeb91b5569ae063" "f2e4389129dff2bfbeaccbdcd29cb100421ba773d074a4a58f0435a240ae0163" "0193b7439bf3c57c620e1bddc4fee2e49d9370f43ed7473ec1cce590325230d3" "0a9d329b09c79b66357e7b8a7e0cf3e83f3315c10494617bca16ea2eecc729fc" "f3b461e27a713ffce08b5b86278f4099aa723b894f97552a80cd639e907b6d19" "b31be606104b6e03aaecc523f4c718e70bfa527057d47cead3a9c933bc6ff528" "554b323d3062cc1839c36eaa0c68f9cc9564046516f6385638299645dfaa6807" "23ab7c54ba6f9008656aa2813e08ff88a853ff01bebf5e484a7dfe88f4dd6ca7" "b27c3420c7d029dc8605591e4eaf27c0d2621de127f937fc597d15422226204f" "63d8c00f02f75774c643acd67c9601416ee32537f5f100a1238b0a0881e4baf7" "0c66d1d0ed4fcf8c413319af96e1c30c549b5ad4af29377141687491792aa7f8" "ac212dbc56e8a68370cb32451f88e682e34af16fbdb6b31ffc7264e3997ac503" "154bfc972d6df3e619577d352b668430422e37c5dd953109cdceacf198f1db68" "e0c66085db350558f90f676e5a51c825cb1e0622020eeda6c573b07cb8d44be5" "b741b866edeaa1df91a7c0dd8f93108fd94cf54af033b3ae7b19eac783468aba" "e297f54d0dc0575a9271bb0b64dad2c05cff50b510a518f5144925f627bb5832" "ed91d4e59412defda16b551eb705213773531f30eb95b69319ecd142fab118ca" "9b59e147dbbde5e638ea1cde5ec0a358d5f269d27bd2b893a0947c4a867e14c1" "0b2e94037dbb1ff45cc3cd89a07901eeed93849524b574fa8daa79901b2bfdcf" "deb7ae3a735635a85c984ece4ce70317268df6027286998b0ea3d10f00764c9b" "0973b33d2f15e6eaf88400eee3dc8357ad8ae83d2ca43c125339b25850773a70" "80a23d559a5c5343a0882664733fd2c9e039b4dbf398c70c424c8d6858b39fc5" "1342a81078bdac27f80b86807b19cb27addc1f9e4c6a637a505ae3ba4699f777" "b4fd44f653c69fb95d3f34f071b223ae705bb691fb9abaf2ffca3351e92aa374" "8530b2f7b281ea6f263be265dd8c75b502ecd7a30b9a0f28fa9398739e833a35" "ed92c27d2d086496b232617213a4e4a28110bdc0730a9457edf74f81b782c5cf" "1a094b79734450a146b0c43afb6c669045d7a8a5c28bc0210aba28d36f85d86f" "7feeed063855b06836e0262f77f5c6d3f415159a98a9676d549bfeb6c49637c4" "5dc8ea509d300f9b751fce3d94d9885d264c5bda240e90e884d48601955b3f8c" "6bb6984321d4e05441b29ce5a40aea95fe1004dc0907124dbbf50852fd2c22de" "a21e81f5aa6f4ed7d9138bd1e282fdaa96012523978044dcb7d41aa59810e54b" "0c3b1358ea01895e56d1c0193f72559449462e5952bded28c81a8e09b53f103f" "ef04dd1e33f7cbd5aa3187981b18652b8d5ac9e680997b45dc5d00443e6a46e3" "4a91a64af7ff1182ed04f7453bb5a4b0c3d82148d27db699df89a5f1d449e2a4" "c968804189e0fc963c641f5c9ad64bca431d41af2fb7e1d01a2a6666376f819c" "12670281275ea7c1b42d0a548a584e23b9c4e1d2dabb747fd5e2d692bcd0d39b" "1177fe4645eb8db34ee151ce45518e47cc4595c3e72c55dc07df03ab353ad132" "b486c4c5f8597e0b8401bc960cc878151a335d20e052ba3c9da27f12881877a2" "5b20570781c33819c0b4bcb009305dbe5a9ed12fcedca10e29f1703b5b9d3f96" "bf7f4fb05a45eae1a6bc1a009b7731b09260d945ec4c3c4ed7f5da06647a7946" "3c7165ae3c29765af5a544d597be36292df073b7db0e95521492569ca74a7dbe" "1ba61848d0d8c78e037867c26f118875705c20f5ad64949a8cee8c8059e5c50f" "9c72688b960c505a8495585ddd9a9764e991884ebf87c7fbedc3f31851d2add2" "22ce4ee620168dc616e03d7154cd4fdc3d5a0277d038150af7db92d960d7a427" "7c32d87e3f35d93754efca582a250e03a3a4eab13cc2f89239df9e977a47c448" "7bf64a1839bf4dbc61395bd034c21204f652185d17084761a648251041b70233" "74c328628e9ced700633c9740666e56f40da5062d93b99b7f2cabb54d0b9194a" "cc0dbb53a10215b696d391a90de635ba1699072745bf653b53774706999208e3" "1b780020c8fe8c91829c06d2a9d5c7d8a182216e479c5b24e787fb1712ffb176" "0c3ed8e4471b7ee7c5f3c921a6aab46af825354f807e3b6595f3871d840560db" "a4d11382b57e6c08c26db2793670642b1fbb828e642cf41ae58685b4e37aeca9" "7527f3308a83721f9b6d50a36698baaedc79ded9f6d5bd4e9a28a22ab13b3cb1" "53d1bb57dadafbdebb5fbd1a57c2d53d2b4db617f3e0e05849e78a4f78df3a1b" "aa0a998c0aa672156f19a1e1a3fb212cdc10338fb50063332a0df1646eb5dfea" "891debfe489c769383717cc7d0020244a8d62ce6f076b2c42dd1465b7c94204d" "723e48296d0fc6e030c7306c740c42685d672fd22337bc84392a1cf92064788a" "4597d1e9bbf1db2c11d7cf9a70204fa42ffc603a2ba5d80c504ca07b3e903770" "bbb4a4d39ed6551f887b7a3b4b84d41a3377535ccccf901a3c08c7317fad7008" "5715d3b4b071d33af95e9ded99a450aad674e308abb06442a094652a33507cd2" "c5d320f0b5b354b2be511882fc90def1d32ac5d38cccc8c68eab60a62d1621f2" "f8cf128fa0ef7e61b5546d12bb8ea1584c80ac313db38867b6e774d1d38c73db" default)))
 '(fci-rule-color "#4C566A")
 '(fringe-mode 6 nil (fringe))
 '(jdee-db-active-breakpoint-face-colors (cons "#191C25" "#80A0C2"))
 '(jdee-db-requested-breakpoint-face-colors (cons "#191C25" "#A2BF8A"))
 '(jdee-db-spec-breakpoint-face-colors (cons "#191C25" "#434C5E"))
 '(linum-format " %7d ")
 '(main-line-color1 "#191919")
 '(main-line-color2 "#111111")
 '(package-selected-packages
   (quote
    (yasnippet smex ido-vertical-mode pdf-tools dired-ranger writeroom-mode olivetti melpa-upstream-visit golden-ratio-scroll-screen highlight-current-line org-beautify-theme org-bullets base16-theme nord-theme magit helm doom-themes autothemer)))
 '(powerline-color1 "#191919")
 '(powerline-color2 "#111111")
 '(size-indication-mode t)
 '(vc-annotate-background "#191C25")
 '(vc-annotate-color-map
   (list
    (cons 20 "#A2BF8A")
    (cons 40 "#bac389")
    (cons 60 "#d3c788")
    (cons 80 "#ECCC87")
    (cons 100 "#e3b57e")
    (cons 120 "#da9e75")
    (cons 140 "#D2876D")
    (cons 160 "#c88982")
    (cons 180 "#be8b98")
    (cons 200 "#B58DAE")
    (cons 220 "#b97e97")
    (cons 240 "#bd6f80")
    (cons 260 "#C16069")
    (cons 280 "#a15b66")
    (cons 300 "#825663")
    (cons 320 "#625160")
    (cons 340 "#4C566A")
    (cons 360 "#4C566A")))
 '(vc-annotate-very-old-color nil))
(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 )



emacs.d/.emacs.d/configuration.el
(setq inhibit-startup-message t)
(setq ring-bell-function 'ignore)
(setq make-backup-files nil)
(menu-bar-mode 0)
(tool-bar-mode 0)
(scroll-bar-mode -1)
(blink-cursor-mode -1)
(set-default 'cursor-type 'box)
(size-indication-mode t)
(global-visual-line-mode t)
;(toggle-truncate-lines t)
;(global-hl-line-mode +1)
;(set-face-background 'hl-line "#F13333")

(global-prettify-symbols-mode t)

(setq-default tab-width 2)

(set-language-environment "UTF-8")
(set-default-coding-systems 'utf-8)

(add-hook 'org-mode-hook #'org-bullets-mode)
(add-hook 'org-mode-hook #'org-indent-mode)

(add-to-list 'org-structure-template-alist
	     '("el" "#+BEGIN_SRC emacs-lisp\n?\n#+END_SRC"))

(org-babel-do-load-languages
 'org-babel-load-languages
 '((emacs-lisp . t)
   (python . t)
   (dot . t)
   (gnuplot . t)))

(setq org-ellipsis "⤵")
(setq org-src-tab-acts-natively t)
(setq org-src-fontify-natively t)

(setq org-confirm-babel-evaluate nil)

(setq ido-enable-flex-matching nil)
(setq ido-create-new-buffer 'always)
(setq ido-everywhere t)
(setq ido-vertical-define-keys 'C-n-and-C-p-only)
(ido-mode 1)
(ido-vertical-mode 1)
;(setq ido-file-extensions-order '(".org" ".txt" ".py" ".emacs" ".xml" ".el" ".ini" ".cfg" ".cnf"))

(defun bind-ido-keys ()
  (define-key ido-completion-map (kbd "C-l") 'ido-up-directory))
(add-hook 'ido-setup-hook #'bind-ido-keys)
;https://emacs.stackexchange.com/questions/3729/how-do-i-bind-keys-in-ido

(require 'golden-ratio-scroll-screen)
(global-set-key (kbd "M-v") 'golden-ratio-scroll-screen-down)
(global-set-key (kbd "C-v") 'golden-ratio-scroll-screen-up)

(global-set-key (kbd "C-x C-b") 'eval-buffer)

(require 'helm-config)
;(helm-mode 1)

;(define-key helm-map (kbd "<tab>") 'helm-execute-persistent-action)

(global-set-key (kbd "M-x") 'helm-M-x)
;(global-set-key (kbd "C-x b") 'helm-buffers-list)
;(global-set-key (kbd "C-x r b") 'helm-bookmarks)
;(global-set-key (kbd "M-y") 'helm-show-kill-ring)
;(global-set-key (kbd "C-x C-f") 'helm-find-files)
;(global-set-key (kbd "C-x p") 'package-list-packages)

;(set-default-font "Liberation Mono 18")
;(set-default-font "Fira Code 18")
(set-default-font "Hack Nerd Font 18")

(setq custom-safe-themes t)
(add-to-list 'custom-theme-load-path "~/.emacs.d/themes")
(add-hook 'after-init-hook (lambda () (load-theme 'xresources)))

(defun my/insert-line-before (times)
  "Inserts a newline(s) above the line containing the cursor."
  (interactive "p")
  (save-excursion
    (move-beginning-of-line 1)
    (newline)))

(global-set-key (kbd "C-S-o")
		'my/insert-line-before)

(setq package-archives
     '(("gnu" . "https://elpa.gnu.org/packages/")
       ("marmalade" . "https://marmalade-repo.org/packages/")
       ("melpa" . "https://melpa.org/packages/")))
